compiler/qsc_fir/src/extensions.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use crate::{
+    fir::{CallableDecl, LocalVarId, Package, Pat, PatId, PatKind},
+    ty::Ty,
+};
+use qsc_data_structures::index_map::IndexMap;
+
+/// The index corresponding to an input parameter.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub struct InputParamIndex(usize);
+
+impl From<InputParamIndex> for usize {
+    fn from(value: InputParamIndex) -> Self {
+        value.0
+    }
+}
+
+impl From<usize> for InputParamIndex {
+    fn from(value: usize) -> Self {
+        Self(value)
+    }
+}
+
+/// An input parameter.
+#[derive(Clone, Debug)]
+pub struct InputParam {
+    pub index: InputParamIndex,
+    pub pat: PatId,
+    pub ty: Ty,
+    pub var: Option<LocalVarId>,
+}
+
+pub trait PackageExt {
+    fn derive_callable_input_params(&self, callable: &CallableDecl) -> Vec<InputParam>;
+}
+
+impl PackageExt for Package {
+    fn derive_callable_input_params(&self, callable: &CallableDecl) -> Vec<InputParam> {
```

Comment:
nitpick: this sort of comes down to personal preference, but since this trait is not used to be generic across implementers anywhere and is only expected to be implemented on `Package` you can actually just implement it as a function for that type right here:

```suggestion
impl Package {
    pub fn derive_callable_input_params(&self, callable: &CallableDecl) -> Vec<InputParam> {
```

Rust doesn't require that all `impl` for a type are in the same block, just that they are in the same crate that defines the type. Since you are in the same crate, this change could be done and the trait removed.

Reply:
Making that change.

Reply:
Updated.